Add PodsReady condition if WaitForPodsReady is enabled

mimowo opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This is the first part of implementing the solution designed in: https://github.com/kubernetes-sigs/kueue/pull/433. 
Admission of queued workloads will be blocked until the currently starting workload has the `PodsReady` condition (when `WaitForPodsReady` is configured at the Kueue level).

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Part of: https://github.com/kubernetes-sigs/kueue/issues/349

#### Special notes for your reviewer:

- the current description of the `WaitForPodsReady` field is not fully implemented - there is going to be a follow up
PR to implement blocking of admission

alculquicondor commented on apis/kueue/v1alpha2/workload_types.go:
```diff
@@ -113,6 +113,10 @@ const (
 	// ResourceClaim finished running (failed or succeeded).
 	WorkloadFinished = "Finished"
+
+	// WorkloadPodsReady means that the first batch of Job pods (up to the
+	// specified parallelism and completions) is either ready or succeeded.
```
That's specific to batch.Job, whereas Workload is supposed to be generic.

Maybe you can say that at least `.spec.podSets[*].replicas` Pods are ready or have succeeded.

mimowo replied:
Adjusted

alculquicondor commented on main.go:
```diff
@@ -120,7 +120,7 @@ func main() {
 	// will block until the cert is ready before setting up the controllers.
 	// Controllers who register after manager starts will start directly.
-	go setupControllers(mgr, cCache, queues, certsReady, cfg.ManageJobsWithoutQueueName)
+	go setupControllers(mgr, cCache, queues, certsReady, cfg.ManageJobsWithoutQueueName, waitForPodsReady(&cfg))
```
Before this line grows too much:

```suggestion
	go setupControllers(mgr, cCache, queues, certsReady, cfg)
```

alculquicondor replied:
This was not addressed.

mimowo replied:
Ups, going to address now.

mimowo replied:
Done

alculquicondor commented on apis/config/v1alpha2/configuration_types.go:
```diff
@@ -45,6 +45,19 @@ type Configuration struct {
+	// admitted and do not block admission of other workloads. The PodsReady
+	// condition is only added if this setting is enabled. If unspecified,
+	// it defaults to false.
+	Enable *bool `json:"enable,omitempty"`
```
Because it defaults to false, this should be enough:

```suggestion
	Enable bool `json:"enable,omitempty"`
```

mimowo replied:
Adjusted

alculquicondor commented on pkg/controller/workload/job/job_controller.go:
```diff
@@ -207,10 +219,25 @@ func (r *JobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
+	} else {
+		log.Info("Job is running", "enabled", r.waitForPodsReady, "ready", jobPodsReady(&job))
+		// handle an unsuspended job when waitForPodsReady is enabled
+		if r.waitForPodsReady && jobPodsReady(&job) {
```
Maybe this logic should be in line 249

alculquicondor replied:
Also, we should add the condition with Status=false if we are still waiting.

mimowo replied:
Good idea. It allows to also distinguish two situations: (1) the job controller has not inspected the Job, (2) the job controller already inspected the job but the pods are not ready.

Is it possible that the workload is in PodsReady condition already, but a user increases the level of parallelism. If so, should we keep the condition or recompute it? Currently I recompute.

alculquicondor commented on pkg/controller/workload/job/job_controller.go:
```diff
@@ -207,10 +219,25 @@ func (r *JobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
 		log.V(3).Info("Job is suspended and workload not yet admitted by a clusterQueue, nothing to do")
 		return ctrl.Result{}, nil
+	} else {
+		log.Info("Job is running", "enabled", r.waitForPodsReady, "ready", jobPodsReady(&job))
```
only log when the value changed? Also, at least V(3)

mimowo replied:
Dropped it entirely

alculquicondor commented on pkg/controller/workload/job/job_controller.go:
```diff
@@ -219,11 +246,33 @@ func (r *JobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
+	return parallelism
+}
+
+func podsReadyCount(jobStatus *batchv1.JobStatus) int32 {
```
You can replace this with `pointer.Int32Deref` https://github.com/kubernetes/utils/blob/master/pointer/pointer.go#L86

mimowo replied:
I see, nice

ahg-g replied:
too many short functions :) I would remove this function and just use `Int32Deref` directly in jobsPodReady

mimowo replied:
Indeed, inlined this function.

alculquicondor commented on pkg/controller/workload/job/job_controller.go:
```diff
@@ -219,11 +246,33 @@ func (r *JobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
+func jobPodsReadyExpectedCount(jobSpec *batchv1.JobSpec) int32 {
+	parallelism := *(jobSpec.Parallelism) // parallelism is always set as it is otherwise defaulted by k8s to 1
+	if jobSpec.Completions != nil {
+		if parallelism <= *jobSpec.Completions {
```
Is this possible? If so, we should fix `ConstructWorkloadFor`.

mimowo replied:
Yes it is, fixed `ConstructWorkloadFor` to reuse the `podsCount` function.

alculquicondor commented on pkg/controller/workload/job/job_controller.go:
```diff
@@ -184,7 +196,7 @@ func (r *JobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
 
 	// 4. Handle a not finished job
```
FYI, if a Job finishes fast, we might never add the `PodsReady` condition. Which is fine, but we need to consider that in the rest of the implementation.

alculquicondor commented on pkg/controller/workload/job/job_controller.go:
```diff
@@ -433,27 +482,49 @@ func ConstructWorkloadFor(ctx context.Context, client client.Client,
+	return appendConditionIfNotExists(conds, generateFinishedCondition(jobStatus))
+}
+
+func generateFinishedCondition(jobStatus batchv1.JobConditionType) metav1.Condition {
```
If this function is not used anywhere else, just make it part of appendFinishedConditionIfNotExists

mimowo replied:
Done

alculquicondor commented on pkg/controller/workload/job/job_controller.go:
```diff
@@ -433,27 +482,49 @@ func ConstructWorkloadFor(ctx context.Context, client client.Client,
+	return appendConditionIfNotExists(conds, generatePodsReadyCondition(jobStatus))
+}
+
+func generatePodsReadyCondition(jobStatus *batchv1.JobStatus) metav1.Condition {
```
same suggestion for this condition

mimowo replied:
Done

alculquicondor commented on pkg/controller/workload/job/job_controller.go:
```diff
@@ -433,27 +482,49 @@ func ConstructWorkloadFor(ctx context.Context, client client.Client,
+	}
+}
+
+func appendConditionIfNotExists(conds []metav1.Condition, condition metav1.Condition) ([]metav1.Condition, bool) {
```
you can use `k8s.io/apimachinery/pkg/api/meta.SetStatusCondition`

mimowo replied:
Good idea, used. Also applied it to add the `WorkloadFinished` condition for consistency.

alculquicondor commented on pkg/controller/workload/job/job_controller_test.go:
```diff
@@ -0,0 +1,149 @@
+import (
+	"testing"
+
+	batchv1 "k8s.io/api/batch/v1"
+
+	"sigs.k8s.io/kueue/pkg/util/pointer"
```
one block

mimowo replied:
Actually, verification started to fail once I joined. It seems the newline separates the k8s and kueue packages. An analogous line is present in other go files.

alculquicondor replied:
ah, yes, this is a kueue package.

alculquicondor commented on pkg/controller/workload/job/job_controller_test.go:
```diff
@@ -0,0 +1,149 @@
+
+	for name, tc := range testcases {
+		t.Run(name, func(t *testing.T) {
+			got := jobPodsReady(tc.job)
```
We already have all the coverage in integration.

mimowo replied:
Not everything, for example in integration tests we only test parallelism=completions. Further, the integration tests are much slower, so I would rather decrease the number of scenarios in integration tests than deleting unit tests.

alculquicondor commented on test/integration/controller/job/job_controller_test.go:
```diff
@@ -265,6 +269,104 @@ var _ = ginkgo.Describe("Job controller for workloads with no queue set", func()
+	ginkgo.DescribeTable("Single job at different stages of progress towards completion",
+		func(jobStatus *batchv1.JobStatus, wantCondition *metav1.Condition) {
+			ginkgo.By("Create a cluster queue")
+			defaultFlavor := testing.MakeResourceFlavor("default").Label(labelKey, "default").Obj()
+			gomega.Expect(k8sClient.Create(ctx, defaultFlavor)).Should(gomega.Succeed())
+			clusterQueue := testing.MakeClusterQueue("cluster-queue").
+				Resource(testing.MakeResource(corev1.ResourceCPU).
+					Flavor(testing.MakeFlavor(defaultFlavor.Name, "5").Obj()).
+					Obj()).Obj()
+			gomega.Expect(k8sClient.Create(ctx, clusterQueue)).Should(gomega.Succeed())
```
this shouldn't be needed

mimowo replied:
indeed

alculquicondor commented on test/integration/controller/job/job_controller_test.go:
```diff
@@ -265,6 +269,104 @@ var _ = ginkgo.Describe("Job controller for workloads with no queue set", func()
+			createdJob.Status = *jobStatus
+			gomega.Expect(k8sClient.Status().Update(ctx, createdJob)).Should(gomega.Succeed())
+
+			ginkgo.By("Verify the PodsReady condition is added if exepected")
```
```suggestion
			ginkgo.By("Verify the PodsReady condition is added")
```

mimowo replied:
Done

alculquicondor commented on pkg/controller/workload/job/job_controller.go:
```diff
@@ -433,27 +482,49 @@ func ConstructWorkloadFor(ctx context.Context, client client.Client,
+
+func generatePodsReadyCondition(jobStatus *batchv1.JobStatus) metav1.Condition {
+	podsReadyCount := podsReadyCount(jobStatus)
+	message := fmt.Sprintf("All pods are ready or succeeded. Ready: %v, succeeded: %v.", podsReadyCount, jobStatus.Succeeded)
```
Use a simple message `Pods are ready or succeeded`. We don't want to generate a Workload update for each change in the Job.

mimowo replied:
Done

mimowo commented on Makefile:
```diff
@@ -41,6 +41,7 @@ BUILDER_IMAGE ?= golang:1.18
 ENVTEST_K8S_VERSION = 1.24
 
 INTEGRATION_TARGET ?= ./test/integration/...
+INTEGRATION_EXTRA_OPTS ?=
```
I use it to pass `--focus` to also target a specific test, for example: `INTEGRATION_EXTRA_OPTS=--focus "Single job at different stages of progress towards completion"`

mimowo replied:
Withdrawn from the change for now as I can also use FIt of FDescribeTable . Let me know what you think, we could also have INTEGRATION_FOCUS dedicated param for focus.

ahg-g commented on pkg/controller/workload/job/job_controller.go:
```diff
@@ -219,11 +244,30 @@ func (r *JobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
 	return ctrl.Result{}, nil
 }
 
+func jobPodsReady(job *batchv1.Job) bool {
```
just "podsReady"

mimowo replied:
Done

ahg-g commented on pkg/controller/workload/job/job_controller.go:
```diff
@@ -219,11 +244,30 @@ func (r *JobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
+	return job.Status.Succeeded+podsReadyCount(&job.Status) >= jobPodsReadyExpectedCount(&job.Spec)
+}
+
+func jobPodsReadyExpectedCount(jobSpec *batchv1.JobSpec) int32 {
```
ditto, remove the job prefix

mimowo replied:
Inlined the entire function as it wasn't called from other places

mimowo commented on pkg/controller/workload/job/job_controller.go:
```diff
@@ -207,10 +219,23 @@ func (r *JobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
 		log.V(3).Info("Job is suspended and workload not yet admitted by a clusterQueue, nothing to do")
 		return ctrl.Result{}, nil
+	} else {
+		if r.waitForPodsReady { // handle an unsuspended job when waitForPodsReady is enabled
```
Done

mimowo commented:
/retest

mimowo commented on pkg/controller/workload/job/job_controller.go:
```diff
@@ -207,10 +219,23 @@ func (r *JobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
 		}
 		log.V(3).Info("Job is suspended and workload not yet admitted by a clusterQueue, nothing to do")
 		return ctrl.Result{}, nil
+	} else {
```
I see. I'm wondering if this is an issue for now - before we have preemption? Maybe this could be a dedicated PR once we have preemption.

mimowo replied:
Done, also added an integration test case in job_controller "Job suspended; PodsReady=True before" which simulates the situation.

mimowo commented:
@alculquicondor @ahg-g thanks for reviewing. Addressed all comments, but one question remains open about setting the condition back to false when suspending the job back. Do we currently have a scenario requiring it? PTAL.

alculquicondor commented on test/integration/controller/job/job_controller_test.go:
```diff
@@ -265,6 +269,186 @@ var _ = ginkgo.Describe("Job controller for workloads with no queue set", func()
+			gomega.Expect(k8sClient.Status().Update(ctx, createdJob)).Should(gomega.Succeed())
+
+			ginkgo.By("Verify the PodsReady condition is added")
+			gomega.Eventually(func() error {
```
```suggestion
			gomega.Eventually(func() metav1.Condition {
```

then you can use `Should(gomega.BeComparableTo(wantCondition))`. This uses `cmp.Diff` underneath

mimowo replied:
done

alculquicondor commented on test/integration/controller/job/job_controller_test.go:
```diff
@@ -265,6 +269,186 @@ var _ = ginkgo.Describe("Job controller for workloads with no queue set", func()
+				return k8sClient.Get(ctx, lookupKey, createdWorkload)
+			}, framework.Timeout, framework.Interval).Should(gomega.Succeed())
+			createdWorkload.Spec.Admission = &kueue.Admission{
+				ClusterQueue: kueue.ClusterQueueReference(clusterQueue.Name),
```
```suggestion
				ClusterQueue: kueue.ClusterQueueReference("foo"),
```

No need for an actual CQ object.

mimowo replied:
done

mimowo commented on pkg/controller/workload/job/job_controller.go:
```diff
@@ -219,11 +247,17 @@ func (r *JobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
 
+// podsReady checks if all pods are ready or succeeded
+func podsReady(job *batchv1.Job) bool {
+	podsReadyCount := pointer.Int32Deref(job.Status.Ready, 0)
```
done

mimowo commented on main.go:
```diff
@@ -245,3 +247,7 @@ func apply(configFile string) (ctrl.Options, config.Configuration) {
 }
+
+func waitForPodsReady(cfg *config.Configuration) bool {
+	return cfg.WaitForPodsReady != nil && cfg.WaitForPodsReady.Enable
```
done

alculquicondor commented on test/integration/framework/framework.go:
```diff
@@ -146,6 +146,15 @@ func DeleteWorkload(ctx context.Context, c client.Client, wl *kueue.Workload) er
 	return nil
 }
 
+func GetConditionByType(list []metav1.Condition, cType string) *metav1.Condition {
```
Just use `apimeta.FindStatusCondition`

mimowo replied:
Done

alculquicondor commented on pkg/workload/workload.go:
```diff
@@ -239,6 +239,10 @@ func UpdateStatusIfChanged(ctx context.Context,
 }
 
 func InCondition(w *kueue.Workload, condition string) bool {
+	return InConditionWithStatus(w, condition, metav1.ConditionTrue)
```
followup clean up: use api/meta

mimowo replied:
Done

alculquicondor commented on pkg/controller/workload/job/job_controller.go:
```diff
@@ -170,11 +183,27 @@ func (r *JobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
+
+	// handle a job when waitForPodsReady is enabled
+	if r.waitForPodsReady {
+		log.V(3).Info("Handling an unsuspended job when waitForPodsReady is enabled")
```
V(5) at least.

alculquicondor replied:
But also, you are not actually checking if the job is suspended.

mimowo replied:
Done

alculquicondor commented on pkg/controller/workload/job/job_controller.go:
```diff
@@ -170,11 +183,27 @@ func (r *JobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
+	}
+
+	// handle a job when waitForPodsReady is enabled
+	if r.waitForPodsReady {
```
you are always returning if you enter this branch. We should continue doing the other checks on the Job

mimowo replied:
Fixed

alculquicondor reviewed (commented):
/approve

mimowo commented on pkg/controller/workload/job/job_controller.go:
```diff
@@ -432,29 +465,41 @@ func ConstructWorkloadFor(ctx context.Context, client client.Client,
+func generatePodsReadyCondition(job *batchv1.Job) metav1.Condition {
+	conditionStatus := metav1.ConditionFalse
+	message := "Not all pods are ready or succeeded"
+	if podsReady(job) {
```
I was thinking about it. However, a suspended job does not need to block admission of new workloads as its pods are terminating, so will not deadlock. Also, a suspended job will eventually reach the PodsReady=False condition as its pods terminate (unless unsuspended quickly, but then we can save the two requests). 

Also, there might be a slight penalty to throuthput by an additional request to set `suspend=false` when a new workload is about to get admitted. WDYT?

alculquicondor replied:
>  a suspended job does not need to block admission of new workloads 

That is true, but for that, we should be looking at `Workload.spec.admission != nil`.

> Also, a suspended job will eventually reach the PodsReady=False condition as its pods terminate (unless unsuspended quickly, but then we can save the two requests).

Since we are unsuspending after updating the jobs ready condition, I think what you have is fine.

mimowo replied:
When the pods are finishing, then the Kueue Job reconciler will set `PodsReady=False` in response do decreased Ready count, but indeed it might be lagging a little bit. So what the scenario would be?
- workload A is admitted and the job A is unsuspended and gets all pods running (`PodsReady=True`)
- workload B preempts A, A gets suspended and removed from admitted workloads
- workload B completes very quickly, so that A is still in `PodsReady=True` ready to be re-admitted
- workload A gets readmitted with `PodsReady=True` even though it has a terminating pod now
- workload C gets admitted even though workload A still needs to restart some of its pods - both A and C are starting their pods at the same time.

However, I think the same scenario may theoretically still happen even if we set `PodsReady=False` when suspending. Still, it is unlikely, but it could happen that workload A gets re-admitted within the Job controller single syncJob pass thus, the pods will start to be terminated by invoking https://github.com/kubernetes/kubernetes/blob/0e19bbb91644885a6db38a77ea3d697730269802/pkg/controller/job/job_controller.go#L1416.

Both scenarios are very unlikely and with this feature we don't aim to give a full guarantee of no concurrent pods scheduling - this may still happen in more likely case of pod restart.

mimowo replied:
Btw. I have just realized we don't add the Pod Disruption condition to pods deleted by the Job controller in reaction to suspending a workload. It seems like an omission / bug? I will create a bug if this is the case.

mimowo replied:
Actually, after a second thought, I see that by setting `PodsReady=False` I would eliminate the risk of the scenario described above by being more conservative.

alculquicondor replied:
It's not a "disruption". The user intention (setting `.spec.suspend=True`) is to terminate those pods.

mimowo replied:
Pushed the suggested change PTAL.

alculquicondor replied:
Maybe worth creating an issue for the discussion around job suspension and the disruption condition. Although, one reason why I thought is not a priority for Jobs is because these pods would get the finalizer removed before a condition is added and they wouldn't affect the failure policy.

alculquicondor commented:
/lgtm
/label tide/merge-method-squash

ahg-g commented on test/integration/controller/job/job_controller_test.go:
```diff
@@ -265,6 +268,217 @@ var _ = ginkgo.Describe("Job controller for workloads with no queue set", func()
 	})
 })
 
+var _ = ginkgo.Describe("Job controller when waitForPodsReady enabled", func() {
```
Can we add a test to verify that the condition gets updated for a workload that initially has `spec.Admission` set and then unset (to simulate preemption).

mimowo replied:
Modified the tests. Now I unconditionally admit the workload, but if a test specifies "suspended=true" then Admission is unset and the job is suspended.

ahg-g replied:
sg, we can have a test in the scheduler once preemption is implemented.

mimowo commented on apis/config/v1alpha2/configuration_types.go:
```diff
@@ -45,6 +45,18 @@ type Configuration struct {
 	// InternalCertManagement is configuration for internalCertManagement
 	InternalCertManagement *InternalCertManagement `json:"internalCertManagement,omitempty"`
+
+	// WaitForPodsReady is configuration for waitForPodsReady
```
Extended :)

mimowo commented on pkg/controller/workload/job/job_controller.go:
```diff
@@ -170,21 +183,35 @@ func (r *JobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
+	// handle a job when waitForPodsReady is enabled
+	if r.waitForPodsReady {
+		log.V(5).Info("Handling a job when waitForPodsReady is enabled")
+		condition := generatePodsReadyCondition(&job)
```
I don't see a practical scenario when it matters. Maybe if the request to set admission=nil, but the request to set the suspend=true fails (and is not executed as the job gets quickly re-admitted), but then it seems to make sense to react to the change of the `suspend` field.

alculquicondor replied:
The only scenario where it could matter is when the workload is immediately admitted again and we didn't have time to update the PodsReady condition to false.

But that could still happen in any case, because we can't update the status and spec at the same time. Unless we can in a mutating webhook?

alculquicondor replied:
We can't reset the condition in a mutating webhook either. To be on the safer side, maybe we can set the condition to false as soon as we see `admission=nil`.

alculquicondor commented on pkg/workload/workload.go:
```diff
@@ -239,6 +240,9 @@ func UpdateStatusIfChanged(ctx context.Context,
+	return InConditionWithStatus(w, condition, metav1.ConditionTrue)
+}
+
+func InConditionWithStatus(w *kueue.Workload, condition string, status metav1.ConditionStatus) bool {
```
Don't add this function.

As a follow up: remove `InCondition`

mimowo replied:
https://github.com/kubernetes-sigs/kueue/pull/474

alculquicondor commented:
/lgtm

Leaving the hold to @ahg-g

ahg-g commented:
/hold cancel
/lgtm
/approve
